### PR TITLE
[flutter_tools] catch SocketException writing to ios-deploy stdin

### LINKED ISSUE #139709: [flutter_tools] SocketException Broken pipe in IOSDeployDebugger.detach
On `stable/3.16.2`
Command `flutter run`

```
SocketException: Write failed (OS Error: Broken pipe, errno = 32), port = 0
at _NativeSocket.write(socket_patch.dart:1246)
at _RawSocket.write(socket_patch.dart:2004)
at _Socket._write(socket_patch.dart:2479)
at _SocketStreamConsumer.write(socket_patch.dart:2214)
at _SocketStreamConsumer.addStream.<anonymous closure>(socket_patch.dart:2168)
at _rootRunUnary(zone.dart:1407)
at _CustomZone.runUnary(zone.dart:1308)
at _CustomZone.runUnaryGuarded(zone.dart:1217)
at _BufferingStreamSubscription._sendData(stream_impl.dart:339)
at _BufferingStreamSubscription._add(stream_impl.dart:271)
at _SyncStreamControllerDispatch._sendData(stream_controller.dart:784)
at _StreamController._add(stream_controller.dart:658)
at _StreamController.add(stream_controller.dart:606)
at _StreamSinkImpl.add(io_sink.dart:154)
at _IOSinkImpl.write(io_sink.dart:287)
at _IOSinkImpl.writeln(io_sink.dart:307)
at _Socket.writeln(socket_patch.dart:2314)
at _StdSink.writeln(stdio.dart:342)
at IOSDeployDebugger.detach(ios_deploy.dart:619)
at IOSDeviceLogReader.dispose(devices.dart:1440)
at IOSDevice.dispose(devices.dart:991)
at HotRunner.cleanupAtFinish(run_hot.dart:1171)
at ResidentRunner.waitForAppToFinish(resident_runner.dart:1483)
at <asynchronous gap>(async)
at HotRunner.attach(run_hot.dart:353)
at <asynchronous gap>(async)
at RunCommand.runCommand(run.dart:745)
at <asynchronous gap>(async)
at FlutterCommand.run.<anonymous closure>(flutter_command.dart:1350)
at <asynchronous gap>(async)
at AppContext.run.<anonymous closure>(context.dart:150)
at <asynchronous gap>(async)
at CommandRunner.runCommand(command_runner.dart:212)
at <asynchronous gap>(async)
at FlutterCommandRunner.runCommand.<anonymous closure>(flutter_command_runner.dart:348)
at <asynchronous gap>(async)
at AppContext.run.<anonymous closure>(context.dart:150)
at <asynchronous gap>(async)
at FlutterCommandRunner.runCommand(flutter_command_runner.dart:294)
at <asynchronous gap>(async)
at run.<anonymous closure>.<anonymous closure>(runner.dart:112)
at <asynchronous gap>(async)
at AppContext.run.<anonymous closure>(context.dart:150)
at <asynchronous gap>(async)
at main(executable.dart:90)
at <asynchronous gap>(async)
```

### PR DESCRIPTION
**Pull request by christopherfujino**
Fixes https://github.com/flutter/flutter/issues/139709

This adds a static helper method `ProcessUtils.writelnToStdinGuarded()`, which will asynchronously write to a sub-process's STDIN `IOSink` and catch errors.

In talking with Brian, it sounds like this is the best and most reliable way to catch `SocketException`s during these writes *to sub-process file descriptors* specifically (with a "real" hard drive file, the future returned by `.flush()` should complete with the write error).

Also, as I note in the dartdoc to `writelnToStdinGuarded()`, the behavior seems to be different between macOS and linux.

Moving forward, in any place where we want to catch exceptions writing to STDIN, we will want to use this new helper.

### CODE REVIEW

File: packages/flutter_tools/lib/src/ios/ios_deploy.dart [hunk 1]
```diff
@@ -598,29 +598,66 @@ class IOSDeployDebugger {
     if (!debuggerAttached) {
       return;
     }
-    try {
-      // Stop the app, which will prompt the backtrace to be printed for all threads in the stdoutSubscription handler.
-      _iosDeployProcess?.stdin.writeln(_signalStop);
-    } on SocketException catch (error) {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/ios/ios_deploy.dart:604]: This would never catch because the exception would be thrown in a different zone

File: packages/flutter_tools/lib/src/ios/ios_deploy.dart [hunk 2]
```diff
@@ -598,29 +598,66 @@ class IOSDeployDebugger {
...
+    if (_stdinWriteFuture != null) {
+      _stdinWriteFuture = _stdinWriteFuture!.then<void>((_) => writeln());
+    } else {
+      _stdinWriteFuture = writeln();
     }
+
+    return _stdinWriteFuture;
+  }
+
+  Future<void> detach() async {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/ios/ios_deploy.dart:649]: making this async is mostly for the benefit of the unit tests. The two non-test calls to this method do not await the future.

File: packages/flutter_tools/lib/src/ios/ios_deploy.dart [hunk 3]
```diff
@@ -598,29 +598,66 @@ class IOSDeployDebugger {
...
+  ///
+  /// No-op if [_iosDeployProcess] is null.
+  ///
+  /// This write will not happen until the flush of any previous writes have
+  /// completed, because calling [IOSink.flush()] before a previous flush has
+  /// completed will throw a [StateError].
+  ///
+  /// This method needs to keep track of the [_stdinWriteFuture] from previous
+  /// calls because the future returned by [detach] is not always await-ed.
+  Future<void> stdinWriteln(String line, {required void Function(Object, StackTrace) onError}) async {
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/ios/ios_deploy.dart:626]: Will the next place `writelnToStdinGuarded` is adopted will also need this logic?  At which time this helper would move into `process.dart`?  Or is this logic special to this `detach`?

Comment (jmagman) [reply packages/flutter_tools/lib/src/ios/ios_deploy.dart:626]: I'm not a `Zone` wizard or even novice, but it seems like scheduling events in a queue so that `writelnToStdinGuarded` isn't called until the flush happens is exactly what a zone is good for?  However, my brain hurts.

File: packages/flutter_tools/lib/src/base/process.dart [hunk 4]
```diff
@@ -233,6 +233,40 @@ abstract class ProcessUtils {
...
+  }) async {
+    final Completer<void> completer = Completer<void>();
+
+    void writeFlushAndComplete() {
+      stdin.writeln(line);
+      stdin.flush().whenComplete(() => completer.complete());
+    }
+
+    runZonedGuarded(writeFlushAndComplete, onError);
+    return completer.future;
```
[the remark this answers is not in the document]
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/base/process.dart:268]: I suppose the `writeln()` *could* error for some other reason, in which case we'd hang forever. Let me take a look at fixing that....

File: packages/flutter_tools/lib/src/ios/ios_deploy.dart (hunk 3 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/ios_deploy.dart:626]: > I'm not a Zone wizard or even novice, but it seems like scheduling events in a queue so that writelnToStdinGuarded isn't called until the flush happens is exactly what a zone is good for?

Are you making a specific recommendation here? If so, can you ELI5? :)

File: packages/flutter_tools/lib/src/base/process.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/base/process.dart:268]: I tried calling `stdin.done.whenComplete(() => completer.complete())` but it turns out that is a DIFFERENT future from the one `.flush()` relies on

[the remark this answers is not in the document]
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/base/process.dart:268]: updated this to ensure we've completed if `runZonedGuarded` caught an error.

**Review by jmagman — APPROVED** [review]

File: packages/flutter_tools/lib/src/ios/ios_deploy.dart (hunk 3 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/ios_deploy.dart:626]: followed up offline, we explored different options but couldn't come up with something that worked.
